File: Solution Segment Tree.cpp

#include <vector>
using namespace std;

class SegmentTree {
    
    int sizeInputWithOffsetForNegativeIntegers{};
    vector<int> segments;

public:
    SegmentTree(int maxAbsoluteValue) {
        sizeInputWithOffsetForNegativeIntegers = 2 * maxAbsoluteValue + 1;
        segments.resize(2 * sizeInputWithOffsetForNegativeIntegers);
    }

    void update(int index, int value) {
        index += sizeInputWithOffsetForNegativeIntegers;
        segments[index] += value;
        while (index > 1) {
            index /= 2;
            segments[index] = segments[2 * index] + segments[2 * index + 1];
        }
    }

    int rangeSum(int left, int right) {
        int sum = 0;
        left += sizeInputWithOffsetForNegativeIntegers;
        right += sizeInputWithOffsetForNegativeIntegers;

        while (left < right) {
            if (left % 2 == 1) {
                sum += segments[left++];
            }
            if (right % 2 == 1) {
                sum += segments[--right];
            }
            left /= 2;
            right /= 2;
        }
        return sum;
    }
};

class Solution {
    
    inline static const int MAX_ABSOLUTE_VALUE = pow(10, 4);

public:
    vector<int> countSmaller(vector<int>& nums) {
        vector<int> resultCountSmaller;
        SegmentTree segmentTree(MAX_ABSOLUTE_VALUE);

        for (int i = nums.size() - 1; i >= 0; --i) {
            int frequency = segmentTree.rangeSum(0, nums[i] + MAX_ABSOLUTE_VALUE);
            resultCountSmaller.push_back(frequency);
            segmentTree.update(nums[i] + MAX_ABSOLUTE_VALUE, 1);
        }

        reverse(resultCountSmaller.begin(), resultCountSmaller.end());
        return resultCountSmaller;
    }
};

File: Solution Segment Tree.js

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var countSmaller = function (nums) {
    const MAX_ABSOLUTE_VALUE = Math.pow(10, 4);
    const segmentTree = new SegmentTree(MAX_ABSOLUTE_VALUE);
    let resultCountSmaller = [];

    for (let i = nums.length - 1; i >= 0; --i) {
        let frequency = segmentTree.rangeSum(0, nums[i] + MAX_ABSOLUTE_VALUE);
        resultCountSmaller.push(frequency);
        segmentTree.update(nums[i] + MAX_ABSOLUTE_VALUE, 1);
    }

    resultCountSmaller.reverse();
    return resultCountSmaller;
};

class SegmentTree {

    constructor(maxAbsoluteValue) {
        this.sizeInputWithOffsetForNegativeIntegers = 2 * maxAbsoluteValue + 1;
        this.segments = new Array(2 * this.sizeInputWithOffsetForNegativeIntegers).fill(0);
    }

    update(index, value) {
        index += this.sizeInputWithOffsetForNegativeIntegers;
        this.segments[index] += value;
        while (index > 1) {
            index = Math.floor(index / 2);
            this.segments[index] = this.segments[2 * index] + this.segments[2 * index + 1];
        }
    }

    rangeSum(left, right) {
        let sum = 0;
        left += this.sizeInputWithOffsetForNegativeIntegers;
        right += this.sizeInputWithOffsetForNegativeIntegers;

        while (left < right) {
            if (left % 2 === 1) {
                sum += this.segments[left++];
            }
            if (right % 2 === 1) {
                sum += this.segments[--right];
            }
            left = Math.floor(left / 2);
            right = Math.floor(right / 2);
        }
        return sum;
    }
}
